libdrgn/elf_symtab.c

```diff
@@ -46,6 +58,98 @@ static struct drgn_error *find_elf_file_symtab(struct drgn_elf_file *file,
 	return NULL;
 }
 
+#ifdef WITH_LZMA
+static struct drgn_error *
+drgn_error_lzma(lzma_ret code)
+{
+	switch (code) {
+		case LZMA_MEM_ERROR:
+			return &drgn_enomem;
+		case LZMA_OPTIONS_ERROR:
+			return drgn_error_format(DRGN_ERROR_INVALID_ARGUMENT,
+						 "lzma: invalid options");
+		case LZMA_FORMAT_ERROR:
+		case LZMA_DATA_ERROR:
+		case LZMA_BUF_ERROR:
+			return drgn_error_format(DRGN_ERROR_INVALID_ARGUMENT,
+						 "lzma: format error (%d)", code);
+		default:
+			return drgn_error_format(DRGN_ERROR_OTHER,
+						 "lzma: unknown error (%d)", code);
+	}
+}
+
+static struct drgn_error *
+load_gnu_debugdata_file(struct drgn_module *module, Elf_Scn *gnu_debugdata_scn)
+{
+	if (module->gnu_debugdata_file)
+		return NULL;
+
+	Elf_Data *gnu_debugdata_data;
+	struct drgn_error *err;
+	err = read_elf_section(gnu_debugdata_scn, &gnu_debugdata_data);
+	if (err)
+		return err;
+
+	_cleanup_(lzma_end) lzma_stream stream = LZMA_STREAM_INIT;
+	lzma_ret ret = lzma_stream_decoder(&stream, UINT64_MAX, 0);
+	if (ret != LZMA_OK)
+		return drgn_error_lzma(ret);
+
+	stream.next_in = gnu_debugdata_data->d_buf;
+	stream.avail_in = gnu_debugdata_data->d_size;
+
+	size_t datasize = 2 * gnu_debugdata_data->d_size;
```

Comment:
I assume this is just a heuristic for a good initial size to start with? Please add a comment to that effect.

Also, it's probably not possible for this to overflow in practice, but I'd rather be paranoid. Maybe something like:

```suggestion
	size_t datasize;
	if (__builtin_mul_overflow(gnu_debugdata_data->size, 2U, &datasize)
		datasize = gnu_debugdata_data->size;
```

Also, `capacity` might be a better name.

Reply:
Yeah, it's just a heuristic guess. And honestly, I'll happily get rid of the guess and just start with the same size of the input, and then all the overflow checks can be done in the same place within the loop.